vm: arithmetic overflow panic in mremap (alloc_free_region_exact) with large size

### Describe the bug

The kernel panics with "attempt to add with overflow" when the mremap syscall is invoked with a crafted, extremely large new_size. The issue is located in VmarInner::alloc_free_region_exact within the VMAR implementation.

# https://github.com/asterinas/asterinas/blob/184e803869d93f7793d33dabcab508f0bec8b5e9/kernel/src/vm/vmar/vmar_impls/mod.rs#L258

### To Reproduce

1. Compile the program and run.

```
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/mman.h>
#include <stdint.h>
#include <errno.h>

int main() {
    // This PoC attempts to trigger an arithmetic overflow in the kernel's remap function
    // The overflow occurs when old_addr + new_size exceeds the maximum address value
    // We'll try to get a high address and then remap with a large size
    
    size_t size = 4096; // Start with a small page
    void *addr;
    
    // Try to allocate memory at high addresses
    // On some systems, we can use MAP_32BIT or other flags to influence placement
    addr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (addr == MAP_FAILED) {
        perror("mmap");
        return 1;
    }
    
    printf("Initial allocation at: %p\n", addr);
    
    // Try to remap with a size that could cause overflow
    // We'll use a very large size value
    // The kernel will check old_addr + new_size for overflow
    size_t new_size = SIZE_MAX - (uintptr_t)addr + 1;
    
    printf("Attempting mremap with size: 0x%lx\n", new_size);
    printf("This would cause overflow in: %p + 0x%lx\n", addr, new_size);
    
    // This should trigger the potential_arithmetic_overflow panic in the kernel
    void *new_addr = mremap(addr, size, new_size, MREMAP_MAYMOVE);
    
    if (new_addr == MAP_FAILED) {
        perror("mremap failed (expected)");
        printf("The kernel likely detected the overflow and returned an error\n");
    } else {
        printf("mremap succeeded unexpectedly at: %p\n", new_addr);
    }
    
    // Clean up
    munmap(addr, size);
    
    return 0;
}
```

### Environment

- Official docker asterinas/asterinas:0.17.0-20251217
- Asterinas version: main [Add unit tests for DMA `alloc` and `alloc_uninit()` · asterinas/asterinas@8ace769](https://github.com/asterinas/asterinas/commit/8ace76978eadd74dd2952866c269ac88b41c577b)

### Logs

```
~ # ./ex1204__asterinas__kernel__src_
_vm__vmar__vmar_impls__remap.rs_105_34 
Initial allocation at: 0x7ffffff2b000
Attempting mremap with size: 0xffff8000000d5000
This would cause overflow in: 0x7ffffff2b000 + 0xffff8000000d5000
[    26.730] ERROR: Uncaught panic:
        attempt to add with overflow
        at /root/asterinas/kernel/src/vm/vmar/vmar_impls/mod.rs:258
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicIdSet { bits: [1], phantom: PhantomData<ostd::cpu::id::CpuId> }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31488980 }, fair: FairAttr { weight: 1024, pending_weight: 0, vruntime: 352426967 } } })
Printing stack trace:
   1: fn 0xffffffff88970160 - pc 0xffffffff889701ab / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6f770;

   2: fn 0xffffffff882187c0 - pc 0xffffffff88218fbd / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6f790;

   3: fn 0xffffffff8821a780 - pc 0xffffffff8821a78a / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fb60;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fb70;

   5: fn 0xffffffff88ae4210 - pc 0xffffffff88ae4256 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fb80;

   6: fn 0xffffffff88aee530 - pc 0xffffffff88aee555 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fbc0;

   7: fn 0xffffffff881b87d0 - pc 0xffffffff881b8871 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fbd0;

   8: fn 0xffffffff881b41a0 - pc 0xffffffff881b4cc3 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd6fc90;

   9: fn 0xffffffff885f7ad0 - pc 0xffffffff885f8225 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd70ca0;

  10: fn 0xffffffff885f78a0 - pc 0xffffffff885f79eb / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd70fa0;

  11: fn 0xffffffff885f89f0 - pc 0xffffffff88600e53 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd710f0;

  12: fn 0xffffffff885f6a50 - pc 0xffffffff885f6aea / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd844f0;

  13: fn 0xffffffff8807e920 - pc 0xffffffff8807f1d1 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84650;

  14: fn 0xffffffff8807f390 - pc 0xffffffff8807f3ee / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84b40;

  15: fn 0xffffffff886617f0 - pc 0xffffffff8866181e / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84cf0;

  16: fn 0xffffffff88661720 - pc 0xffffffff88661766 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84d20;

  17: fn 0xffffffff885c94f0 - pc 0xffffffff885c951c / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84d90;

  18: fn 0xffffffff8813b520 - pc 0xffffffff8813b55f / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84dd0;

  19: fn 0xffffffff8807f350 - pc 0xffffffff8807f377 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84f40;

  20: fn 0xffffffff8805f290 - pc 0xffffffff8805f2ad / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84f80;

  21: fn 0xffffffff88945e40 - pc 0xffffffff88945e65 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84fa0;

  22: fn 0xffffffff889631a0 - pc 0xffffffff88963224 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd84fd0;

  23: fn 0xffffffff8896df2d - pc 0xffffffff8896df36 / registers:

     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd85000;


     rax 0xffffdfffffd6f688; rdx 0xffffffff889a18a2; rcx  0x10000020100a6c5; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bc93f8; rbp                0x0; rsp 0xffffdfffffd85008;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ostd::panic::print_stack_trace
          at /root/asterinas/ostd/src/panic.rs:113
(  2) aster_kernel::thread::oops::panic_handler
          at /root/asterinas/kernel/src/thread/oops.rs:130
(  3) __ostd_panic_handler
          at ??:?
(  4) __rustc::rust_begin_unwind
          at 5hffciyz3d33t9gfe7c8dzufi:?
(  5) core::panicking::panic_fmt
          at ??:?
(  6) core::panicking::panic_const::panic_const_add_overflow
          at ??:?
(  7) <aster_kernel::vm::vmar::vmar_impls::VmarInner>::alloc_free_region_exact
          at 3l0bfq47xxjnzhego6qg8sutw:?
(  8) <aster_kernel::vm::vmar::vmar_impls::Vmar>::remap
          at /root/asterinas/kernel/src/vm/vmar/vmar_impls/remap.rs:138
(  9) aster_kernel::syscall::mremap::do_sys_mremap
          at /root/asterinas/kernel/src/syscall/mremap.rs:68
( 10) aster_kernel::syscall::mremap::sys_mremap
          at /root/asterinas/kernel/src/syscall/mremap.rs:18
( 11) aster_kernel::syscall::arch::syscall_dispatch
          at /root/asterinas/kernel/src/syscall/mod.rs:244
( 12) aster_kernel::syscall::handle_syscall
          at /root/asterinas/kernel/src/syscall/mod.rs:374
( 13) aster_kernel::thread::task::create_new_user_task::{closure#0}
          at /root/asterinas/kernel/src/thread/task.rs:95
( 14) aster_kernel::thread::task::create_new_user_task::{closure#1}
          at /root/asterinas/kernel/src/thread/task.rs:126
( 15) unwinding::panicking::catch_unwind::do_call::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:56
( 16) unwinding::panicking::catch_unwind::<unwinding::panic::RustPanic, (), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:42
( 17) unwinding::panic::catch_unwind::<(), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panic.rs:87
( 18) aster_kernel::thread::oops::catch_panics_as_oops::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/asterinas/kernel/src/thread/oops.rs:55
( 19) aster_kernel::thread::task::create_new_user_task::{closure#2}
          at /root/asterinas/kernel/src/thread/task.rs:131
( 20) <aster_kernel::thread::task::create_new_user_task::{closure#2} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 21) <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2206
( 22) kernel_task_entry
          at /root/asterinas/ostd/src/task/mod.rs:196
( 23) kernel_task_entry_wrapper
          at ??:?
make: *** [Makefile:309: run_kernel] Error 1
```

--- a/kernel/src/vm/vmar/vmar_impls/remap.rs
+++ b/kernel/src/vm/vmar/vmar_impls/remap.rs
@@ -134,9 +134,10 @@ impl Vmar {
             debug_assert!(new_size > old_size);
 
             // Fast path: expand the old mapping in place to the new size
-            if inner
-                .alloc_free_region_exact(old_range.end, new_size - old_size)
-                .is_ok()
+            if is_userspace_vaddr_range(old_addr, new_size)
+                && inner
+                    .alloc_free_region_exact(old_range.end, new_size - old_size)
+                    .is_ok()
             {
                 let old_mapping_addr = inner.check_lies_in_single_mapping(old_addr, old_size)?;
                 let old_mapping = inner.remove(&old_mapping_addr).unwrap();

